### DISCUSSION TITLE
Viewport

### DISCUSSION THREAD

**Discussion by queryFF** [Mapsui/Mapsui#2287] [Q&A] [answered]
Another basic question (please feel free to direct me to a document somewhere if I missed it).  The Viewport has:

CenterX : longitude in world coordinates
CenterY: latitude in world coordinates
Resolution: ?? (a single double value) ??
ScreenToWorld(): convert screen position to world coordinates
ScreenToWorldXY(): seems to be the same as ScreenToWorld
Height: Height of the viewable area in screen units
Width: Width of the viewable area in screen units

mapControl.Map.Navigator.Viewport.ToExtent(): returns the min/max X/Y values, in world coordinates, of the portion of the map displayed in the Viewport.

It's possible to zoom in to the point that the entire map is shown in just a portion of the Viewport and there is whitespace around the displayed map.

I'd like to determine the extent of the map area displayed in the Viewport.  For each corner of the displayed area of the map, I'd like to calculate 1) the corner is shown at (x,y) screen pixel, 2) which corresponds to world (x,y) coordinates, 3) which corresponds to (longitude,latitude).

I would greatly appreciate any pointers as to how to do this.

**Comment by pauldendulk** [comment]
There is some API documentation of Viewport:
https://mapsui.com/api/Mapsui.Viewport.html
And the ViewportExtensions for the ScreenToWorld(XY):
https://mapsui.com/api/Mapsui.Extensions.ViewportExtensions.html
Here is page specific about what Resolution means:
https://mapsui.com/documentation/resolution.html

You need to know there are three types of coordinates you usually deal with
- Screen coordinates, pixels, or more accurate [device independent units](https://mapsui.com/documentation/device-independent-units.html). You could use `mapControl.Map.Navigator.Viewport.ToExtent():`
- The coordinate system of the map. If you use the openstreetmap tile layers that is spherical mercator. You can get those by converting with TheScreenToWorld extensions.
- Longitude, Latitude coordinates (WGS84, EPSG:4326). You can use SphercialMercator.ToLonLat to convert the map coordinates. Read here about projections https://mapsui.com/documentation/projections.html

**Comment by queryFF** [comment, marked as answer]
Thanks Paul!  That was very helpful.  The resolution explanation --> zoom level makes perfect sense.  A quick experiment with Viewport.ToExtent() demonstrates that the world coordinate values will correspond with the size of the Viewport so I should expect values that will fall outside of the visible map when zoomed out to the world view with white space framing the map.

I imagine if the X /Y extent values are outside of the range +/- 4007517/2, I can just use that value and WorldToScreen to get the actual screen diu coordinates of the map edges.

Separately, a quick note from my looking around: you should be able to remove the IDisposable declaration from GeometryFeature since it extends BaseFeature and is declared there.

Thanks again for the help.

**Reply by pauldendulk** [reply]
> I imagine if the X /Y extent values are outside of the range +/- 4007517/2, I can just use that value and WorldToScreen to get the actual screen diu coordinates of the map edges.

I have never taken into account such extreme values, but I expect it to work for WorldToScreen (and back), for SphericalMercator you will probably get non-valid coordinates.

**Reply by pauldendulk** [reply]
I am not sure what you intend to do, but we have code to Fit or Fill an extent in a viewport, perhaps this is helpful. You could take a look at the implementation of Navigator.ZoomToBox.

**Reply by queryFF** [reply]
Will do.  Thanks!  Just kicking the tires right now.  You guys have done a nice job with this.
